File: zig/src/rpi_info.zig
const std = @import("std");
const posix = std.posix;
const linux = std.os.linux;

const log = std.log.scoped(.rpi_info);

pub const TemperatureType = enum { celsius, fahrenheit };
pub const temperature_type: TemperatureType = .celsius;
pub const refresh_interval_secs: u64 = 5;

fn hasPrefix(s: []const u8, prefix: []const u8) bool {
    return std.mem.startsWith(u8, s, prefix);
}

pub fn getIpAddress() []const u8 {
    const StaticBuf = struct {
        var iface_buf: [64]u8 = undefined;
        var ip_buf: [64]u8 = undefined;
    };

    const file = std.fs.openFileAbsolute("/proc/net/route", .{}) catch {
        log.err("failed to open /proc/net/route", .{});
        return "no network";
    };
    defer file.close();

    var buf: [4096]u8 = undefined;
    var reader = file.reader();

    // Skip header line
    _ = reader.readUntilDelimiter(&buf, '\n') catch return "no network";

    var default_iface: ?[]const u8 = null;
    while (reader.readUntilDelimiter(&buf, '\n') catch null) |line| {
        var iter = std.mem.tokenizeAny(u8, line, " \t");
        const iface = iter.next() orelse continue;
        const dest = iter.next() orelse continue;
        if (std.mem.eql(u8, dest, "00000000")) {
            const len = @min(iface.len, StaticBuf.iface_buf.len);
            @memcpy(StaticBuf.iface_buf[0..len], iface[0..len]);
            default_iface = StaticBuf.iface_buf[0..len];
            break;
        }
    }

    const iface_name = default_iface orelse return "no network";

    // Use ioctl to get the IP address via SIOCGIFADDR
    const sock = posix.socket(posix.AF.INET, posix.SOCK.DGRAM, 0) catch return "no network";
    defer posix.close(sock);

    var ifr: linux.ifreq = std.mem.zeroes(linux.ifreq);
    const name_len = @min(iface_name.len, ifr.ifrn.name.len - 1);
    @memcpy(ifr.ifrn.name[0..name_len], iface_name[0..name_len]);
    ifr.ifrn.name[name_len] = 0;

    const SIOCGIFADDR: u32 = 0x8915;
    const result = linux.ioctl(@bitCast(sock), SIOCGIFADDR, @intFromPtr(&ifr));
    if (@as(isize, @bitCast(result)) != 0) {
        return "no network";
    }

    // Extract the IPv4 address from sockaddr_in
    const addr_bytes = ifr.ifru.addr.data[2..6];
    const formatted = std.fmt.bufPrint(&StaticBuf.ip_buf, "{}.{}.{}.{}", .{
        addr_bytes[0], addr_bytes[1], addr_bytes[2], addr_bytes[3],
    }) catch return "no network";
    return formatted;
}

pub fn getRamPercent() u8 {
    const file = std.fs.openFileAbsolute("/proc/meminfo", .{}) catch {
        log.err("failed to open /proc/meminfo", .{});
        return 0;
    };
    defer file.close();

    var buf: [4096]u8 = undefined;
    var reader = file.reader();
    var total: u64 = 0;
    var avail: u64 = 0;

    while (reader.readUntilDelimiter(&buf, '\n') catch null) |line| {
        var iter = std.mem.tokenizeAny(u8, line, " \t");
        const label = iter.next() orelse continue;
        const value_str = iter.next() orelse continue;
        const value = std.fmt.parseInt(u64, value_str, 10) catch continue;

        if (std.mem.eql(u8, label, "MemTotal:")) {
            total = value;
        } else if (std.mem.eql(u8, label, "MemAvailable:")) {
            avail = value;
        }
    }

    if (total == 0) return 0;
    return @intCast((total - avail) * 100 / total);
}

const StatfsResult = struct {
    total_gib: u32,
    used_gib: u32,
};

// Linux statfs64 struct for aarch64
const Statfs = extern struct {
    f_type: i64,
    f_bsize: i64,
    f_blocks: u64,
    f_bfree: u64,
    f_bavail: u64,
    f_files: u64,
    f_ffree: u64,
    f_fsid: [2]i32,
    f_namelen: i64,
    f_frsize: i64,
    f_flags: i64,
    f_spare: [4]i64,
};

fn doStatfs(path: [*:0]const u8) ?StatfsResult {
    var info: Statfs = undefined;
    // Use raw syscall: SYS_statfs on aarch64 is 43
    const ret = linux.syscall2(.statfs, @intFromPtr(path), @intFromPtr(&info));
    if (@as(isize, @bitCast(ret)) != 0) return null;

    const block: u64 = @intCast(info.f_bsize);
    const total_bytes: u64 = block * info.f_blocks;
    const free_bytes: u64 = block * info.f_bfree;
    const used_bytes: u64 = total_bytes - free_bytes;
    return StatfsResult{
        .total_gib = @intCast(total_bytes >> 30),
        .used_gib = @intCast(used_bytes >> 30),
    };
}

fn getSdMemory() StatfsResult {
    return doStatfs("/") orelse blk: {
        log.err("statfs(\"/\") failed", .{});
        break :blk StatfsResult{ .total_gib = 0, .used_gib = 0 };
    };
}

fn getHardDiskMemory() StatfsResult {
    var result = StatfsResult{ .total_gib = 0, .used_gib = 0 };

    const file = std.fs.openFileAbsolute("/proc/mounts", .{}) catch {
        log.err("failed to open /proc/mounts", .{});
        return result;
    };
    defer file.close();

    var buf: [4096]u8 = undefined;
    var reader = file.reader();

    while (reader.readUntilDelimiter(&buf, '\n') catch null) |line| {
        var iter = std.mem.tokenizeAny(u8, line, " \t");
        const device = iter.next() orelse continue;
        const mountpoint = iter.next() orelse continue;

        if (hasPrefix(device, "/dev/sda") or hasPrefix(device, "/dev/nvme")) {
            // Need null-terminated string for statfs syscall
            var mp_buf: [256]u8 = undefined;
            if (mountpoint.len >= mp_buf.len) continue;
            @memcpy(mp_buf[0..mountpoint.len], mountpoint);
            mp_buf[mountpoint.len] = 0;
            const mp_z: [*:0]const u8 = @ptrCast(mp_buf[0..mountpoint.len :0]);

            if (doStatfs(mp_z)) |info| {
                result.total_gib += info.total_gib;
                result.used_gib += info.used_gib;
            }
        }
    }

    return result;
}

pub fn getDiskPercent() u8 {
    const sd = getSdMemory();
    const disk = getHardDiskMemory();

    const total = sd.total_gib + disk.total_gib;
    const used = sd.used_gib + disk.used_gib;

    if (total == 0) return 0;
    const pct = used * 100 / total;
    return @intCast(if (pct > 100) 100 else pct);
}

pub fn getTemperature() u8 {
    const file = std.fs.openFileAbsolute("/sys/class/thermal/thermal_zone0/temp", .{}) catch {
        log.err("failed to open thermal_zone0", .{});
        return 0;
    };
    defer file.close();

    var buf: [16]u8 = undefined;
    const bytes_read = file.read(&buf) catch return 0;
    if (bytes_read == 0) return 0;

    const trimmed = std.mem.trimRight(u8, buf[0..bytes_read], " \t\n\r");
    const millideg = std.fmt.parseInt(u32, trimmed, 10) catch return 0;

    const celsius: u8 = @intCast(millideg / 1000);
    if (temperature_type == .fahrenheit) {
        return @intCast(@as(u32, celsius) * 9 / 5 + 32);
    }
    return celsius;
}

const CpuState = struct {
    prev_idle: u64 = 0,
    prev_total: u64 = 0,
    initialized: bool = false,
};

var cpu_state = CpuState{};

fn readCpuStat() ?struct { idle: u64, total: u64 } {
    const file = std.fs.openFileAbsolute("/proc/stat", .{}) catch {
        log.err("failed to open /proc/stat", .{});
        return null;
    };
    defer file.close();

    var buf: [512]u8 = undefined;
    const bytes_read = file.read(&buf) catch return null;
    if (bytes_read == 0) return null;

    const line_end = std.mem.indexOfScalar(u8, buf[0..bytes_read], '\n') orelse bytes_read;
    var iter = std.mem.tokenizeAny(u8, buf[0..line_end], " \t");

    // Skip "cpu" label
    _ = iter.next() orelse return null;

    var values: [8]u64 = undefined;
    for (&values) |*v| {
        const tok = iter.next() orelse return null;
        v.* = std.fmt.parseInt(u64, tok, 10) catch return null;
    }

    const user = values[0];
    const nice = values[1];
    const system = values[2];
    const idle_val = values[3];
    const iowait = values[4];
    const irq = values[5];
    const softirq = values[6];
    const steal = values[7];

    return .{
        .idle = idle_val + iowait,
        .total = user + nice + system + idle_val + iowait + irq + softirq + steal,
    };
}

pub fn getCpuPercent() u8 {
    if (!cpu_state.initialized) {
        const stats = readCpuStat() orelse return 0;
        cpu_state.prev_idle = stats.idle;
        cpu_state.prev_total = stats.total;
        std.time.sleep(100 * std.time.ns_per_ms);
        cpu_state.initialized = true;
    }

    const stats = readCpuStat() orelse return 0;

    const diff_idle = stats.idle - cpu_state.prev_idle;
    const diff_total = stats.total - cpu_state.prev_total;
    cpu_state.prev_idle = stats.idle;
    cpu_state.prev_total = stats.total;

    if (diff_total == 0) return 0;
    return @intCast((100 * (diff_total - diff_idle) + diff_total / 2) / diff_total);
}

pub fn getHostname() []const u8 {
    const HostBuf = struct {
        var buf: [65]u8 = undefined;
    };
    var uts: linux.utsname = undefined;
    const ret = linux.uname(&uts);
    if (@as(isize, @bitCast(ret)) != 0) {
        return "unknown";
    }
    const nodename: [*:0]const u8 = @ptrCast(&uts.nodename);
    const len = std.mem.len(nodename);
    const copy_len = @min(len, HostBuf.buf.len - 1);
    @memcpy(HostBuf.buf[0..copy_len], nodename[0..copy_len]);
    HostBuf.buf[copy_len] = 0;
    return HostBuf.buf[0..copy_len];
}

pub fn getDietpiUpdateStatus() i32 {
    std.fs.accessAbsolute("/run/dietpi", .{}) catch return 0;
    std.fs.accessAbsolute("/run/dietpi/.update_available", .{}) catch return 1;
    return 2;
}

pub fn getAptUpdateCount() i32 {
    const file = std.fs.openFileAbsolute("/run/dietpi/.apt_updates", .{}) catch return -1;
    defer file.close();

    var buf: [16]u8 = undefined;
    const bytes_read = file.read(&buf) catch return 0;
    if (bytes_read == 0) return 0;

    const trimmed = std.mem.trimRight(u8, buf[0..bytes_read], " \t\n\r");
    return std.fmt.parseInt(i32, trimmed, 10) catch 0;
}

File: rust/src/rpi_info.rs
//! Raspberry Pi system metrics collection.

use std::fs;
use std::path::Path;

/// Temperature unit configuration.
#[derive(Clone, Copy, PartialEq)]
pub enum TemperatureType { Celsius, Fahrenheit }
pub const TEMPERATURE_TYPE: TemperatureType = TemperatureType::Celsius;

/// Dashboard refresh interval in seconds.
pub const REFRESH_INTERVAL_SECS: u64 = 5;

/// Get the IP address of the default network interface.
pub fn get_ip_address() -> String {
    let route_contents = match fs::read_to_string("/proc/net/route") {
        Ok(c) => c,
        Err(_) => {
            eprintln!("rpiInfo: failed to open /proc/net/route");
            return "no network".to_string();
        }
    };

    let iface_name = route_contents
        .lines()
        .skip(1)
        .find_map(|line| {
            let mut fields = line.split_whitespace();
            let name = fields.next()?;
            if fields.next() == Some("00000000") { Some(name) } else { None }
        });

    let iface_name = match iface_name {
        Some(name) => name,
        None => return "no network".to_string(),
    };

    // Look up the interface's IPv4 address using getifaddrs
    let addrs = match nix::ifaddrs::getifaddrs() {
        Ok(a) => a,
        Err(_) => return "no network".to_string(),
    };

    for ifaddr in addrs {
        if ifaddr.interface_name == iface_name {
            if let Some(sin) = ifaddr.address.as_ref().and_then(|a| a.as_sockaddr_in()) {
                return std::net::Ipv4Addr::from(sin.ip()).to_string();
            }
        }
    }

    "no network".to_string()
}

/// Get RAM usage as a percentage (0-100).
pub fn get_ram_percent() -> u8 {
    let contents = match fs::read_to_string("/proc/meminfo") {
        Ok(c) => c,
        Err(_) => {
            eprintln!("rpiInfo: failed to open /proc/meminfo");
            return 0;
        }
    };

    let mut total: u64 = 0;
    let mut avail: u64 = 0;

    for line in contents.lines() {
        let mut parts = line.split_whitespace();
        let label = parts.next();
        let value = parts.next();
        match (label, value) {
            (Some("MemTotal:"), Some(v)) => total = v.parse().unwrap_or(0),
            (Some("MemAvailable:"), Some(v)) => avail = v.parse().unwrap_or(0),
            _ => {}
        }
    }

    if total == 0 {
        return 0;
    }
    ((total - avail) * 100 / total) as u8
}

fn get_sd_memory() -> (u32, u32) {
    match nix::sys::statfs::statfs("/") {
        Ok(info) => {
            let block = info.block_size() as u64;
            let total = block * info.blocks() as u64;
            let used = total - block * info.blocks_free() as u64;
            ((total >> 30) as u32, (used >> 30) as u32)
        }
        Err(_) => {
            eprintln!("rpiInfo: statfs(\"/\") failed");
            (0, 0)
        }
    }
}

fn get_hard_disk_memory() -> (u32, u32) {
    let mut total_gib: u32 = 0;
    let mut used_gib: u32 = 0;

    let contents = match fs::read_to_string("/proc/mounts") {
        Ok(c) => c,
        Err(_) => {
            eprintln!("rpiInfo: failed to open /proc/mounts");
            return (0, 0);
        }
    };

    for line in contents.lines() {
        let mut parts = line.split_whitespace();
        let (device, mountpoint) = match (parts.next(), parts.next()) {
            (Some(d), Some(m)) => (d, m),
            _ => continue,
        };
        if device.starts_with("/dev/sda") || device.starts_with("/dev/nvme") {
            if let Ok(info) = nix::sys::statfs::statfs(mountpoint) {
                let block = info.block_size() as u64;
                let total = block * info.blocks() as u64;
                let used = total - block * info.blocks_free() as u64;
                total_gib += (total >> 30) as u32;
                used_gib += (used >> 30) as u32;
            }
        }
    }

    (total_gib, used_gib)
}

/// Get total disk usage (SD + hard disks) as a percentage (0-100).
pub fn get_disk_percent() -> u8 {
    let (sd_total, sd_used) = get_sd_memory();
    let (disk_total, disk_used) = get_hard_disk_memory();

    let total = sd_total + disk_total;
    let used = sd_used + disk_used;

    if total == 0 {
        return 0;
    }
    (used * 100 / total).min(100) as u8
}

/// Get CPU temperature in the configured unit (Celsius or Fahrenheit).
pub fn get_temperature() -> u8 {
    let contents = match fs::read_to_string("/sys/class/thermal/thermal_zone0/temp") {
        Ok(c) => c,
        Err(_) => {
            eprintln!("rpiInfo: failed to open thermal_zone0");
            return 0;
        }
    };

    let millideg: u32 = match contents.trim().parse() {
        Ok(v) => v,
        Err(_) => return 0,
    };

    let celsius = millideg / 1000;
    match TEMPERATURE_TYPE {
        TemperatureType::Fahrenheit => (celsius * 9 / 5 + 32) as u8,
        TemperatureType::Celsius => celsius as u8,
    }
}

fn read_cpu_stat() -> Option<(u64, u64)> {
    let contents = match fs::read_to_string("/proc/stat") {
        Ok(c) => c,
        Err(_) => {
            eprintln!("rpiInfo: failed to open /proc/stat");
            return None;
        }
    };

    let first_line = contents.lines().next()?;
    let mut parts = first_line.split_whitespace();
    if parts.next() != Some("cpu") {
        return None;
    }

    let mut next_val = || -> Option<u64> { parts.next()?.parse().ok() };
    let user = next_val()?;
    let nice = next_val()?;
    let system = next_val()?;
    let idle_val = next_val()?;
    let iowait = next_val()?;
    let irq = next_val()?;
    let softirq = next_val()?;
    let steal = next_val()?;

    let idle = idle_val + iowait;
    let total = user + nice + system + idle_val + iowait + irq + softirq + steal;
    Some((idle, total))
}

/// CPU usage tracker that maintains state between calls.
pub struct CpuTracker {
    prev_idle: u64,
    prev_total: u64,
    initialized: bool,
}

impl CpuTracker {
    pub fn new() -> Self {
        CpuTracker {
            prev_idle: 0,
            prev_total: 0,
            initialized: false,
        }
    }

    /// Get current CPU usage as a percentage (0-100).
    pub fn get_cpu_percent(&mut self) -> u8 {
        if !self.initialized {
            if let Some((idle, total)) = read_cpu_stat() {
                self.prev_idle = idle;
                self.prev_total = total;
            } else {
                return 0;
            }
            std::thread::sleep(std::time::Duration::from_millis(100));
            self.initialized = true;
        }

        let (idle, total) = match read_cpu_stat() {
            Some(v) => v,
            None => return 0,
        };

        let diff_idle = idle - self.prev_idle;
        let diff_total = total - self.prev_total;
        self.prev_idle = idle;
        self.prev_total = total;

        if diff_total == 0 {
            return 0;
        }
        ((100 * (diff_total - diff_idle) + diff_total / 2) / diff_total) as u8
    }
}

/// Get the system hostname.
pub fn get_hostname() -> String {
    match nix::sys::utsname::uname() {
        Ok(info) => info.nodename().to_string_lossy().into_owned(),
        Err(_) => "unknown".to_string(),
    }
}

/// Check DietPi update status.
/// Returns: 0 = not DietPi, 1 = up to date, 2 = update available.
pub fn get_dietpi_update_status() -> i32 {
    if !Path::new("/run/dietpi").exists() {
        return 0;
    }
    if Path::new("/run/dietpi/.update_available").exists() {
        return 2;
    }
    1
}

/// Get the number of pending APT updates.
/// Returns: -1 on error, 0+ = count.
pub fn get_apt_update_count() -> i32 {
    match fs::read_to_string("/run/dietpi/.apt_updates") {
        Ok(contents) => contents.trim().parse().unwrap_or(0),
        Err(_) => -1,
    }
}
